### PR TITLE
feat(optimizer): support subquery unnest for project set

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Support subquery unnest for project set. We need to ensure that if on condition of the apply operator refer to the table function of ProjectSet,  reject it to unnest.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/output/subquery.yaml [hunk 1]
```diff
@@ -516,8 +516,68 @@
...
+    └─LogicalAgg { group_key: [t.x, t.k], aggs: [sum(Unnest($0))] }
+      └─LogicalJoin { type: LeftOuter, on: IsNotDistinctFrom(t.x, t.x) AND IsNotDistinctFrom(t.k, t.k), output: [t.x, t.k, Unnest($0)] }
+        ├─LogicalAgg { group_key: [t.x, t.k], aggs: [] }
+        │ └─LogicalScan { table: t, columns: [t.x, t.k] }
+        └─LogicalProject { exprs: [t.x, t.k, Unnest($0)] }
+          └─LogicalProjectSet { select_list: [$0, $1, Unnest($0)] }
+            └─LogicalJoin { type: Inner, on: true, output: all }
+              ├─LogicalAgg { group_key: [t.x, t.k], aggs: [] }
+              │ └─LogicalScan { table: t, columns: [t.x, t.k], predicate: (t.k = 1:Int32) }
+              └─LogicalValues { rows: [[]], schema: Schema { fields: [] } }
```
Comment (chenzl25) [inline src/frontend/planner_test/tests/testdata/output/subquery.yaml:578]: We need another rule to remove this cross join (with a simple value as right input). In this way we can support streaming query in the future.

**Review by st1page — APPROVED** [review]
LGTM

File: src/frontend/planner_test/tests/testdata/output/subquery.yaml [hunk 2]
```diff
@@ -516,8 +516,68 @@
...
```
Comment (fuyufjh) [inline src/frontend/planner_test/tests/testdata/output/subquery.yaml:565]: Where is this `Join` from?

File: src/frontend/planner_test/tests/testdata/output/subquery.yaml [hunk 3]
```diff
@@ -516,8 +516,68 @@
...
+    └─LogicalAgg { group_key: [t.x], aggs: [sum(Unnest($0))] }
+      └─LogicalJoin { type: LeftOuter, on: IsNotDistinctFrom(t.x, t.x), output: [t.x, Unnest($0)] }
+        ├─LogicalAgg { group_key: [t.x], aggs: [] }
+        │ └─LogicalScan { table: t, columns: [t.x] }
+        └─LogicalProject { exprs: [t.x, Unnest($0)] }
+          └─LogicalProjectSet { select_list: [$0, Unnest($0)] }
+            └─LogicalJoin { type: Inner, on: true, output: all }
+              ├─LogicalAgg { group_key: [t.x], aggs: [] }
+              │ └─LogicalScan { table: t, columns: [t.x] }
+              └─LogicalValues { rows: [[]], schema: Schema { fields: [] } }
```
Comment (fuyufjh) [inline src/frontend/planner_test/tests/testdata/output/subquery.yaml:535]: Where is this `Join` from?

Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/output/subquery.yaml:535]: It is generated from Apply passing through `select unnest(x)` whose plan is a ProjectSet + a Value. Finally, Apply could be transformed into a join.

**Pull request merged**